File: request-service/internal/requests/service.go
// Package requests implements the purchase-request domain (spec section 10).
package requests

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgtype"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/hmidaayoub/marketplace-microservices/request-service/internal/events"
	"github.com/hmidaayoub/marketplace-microservices/request-service/internal/store"
)

// RequestExistsError refuses a create because the item already has an open request, and
// carries that request so the caller has somewhere to go. It is a value rather than a
// sentinel because the request is the point: "this already exists" without saying what
// leaves the customer with nothing to do about it.
//
// This is the only thing that stops a create. A merely similar name does not: whether a
// close spelling is the same product is a judgement about products, and the service has
// only the spelling to go on, so those are put to the customer as suggestions and
// decided by them. The same name is not a judgement - it is the state the platform
// exists to prevent, and there is nothing for a second request to add.
//
// The request it carries may have no buyers on it - one a seller opened by offering
// against the item, or one everybody has left. That is still the request to join rather
// than one to duplicate: a join is what makes it demand again, and it may already have
// offers waiting on it.
type RequestExistsError struct {
	Existing store.PurchaseRequest
}

func (e *RequestExistsError) Error() string {
	return "an open request already carries this item name"
}

var (
	ErrRequestNotFound    = errors.New("request not found")
	ErrNotParticipant     = errors.New("caller is not a participant in this request")
	ErrAlreadyParticipant = errors.New("caller has already joined this request")
)

// A request describes demand, and demand does not end - it only stops having anyone
// behind it. So there are two statuses and neither is terminal: OPEN while at least one
// customer wants the item, INACTIVE once the last one leaves, and a join makes it OPEN
// again. Both are written by RecalculateDemand from the participant count, which is why
// nothing here sets a status and no API accepts one.
const (
	StatusOpen     = "OPEN"
	StatusInactive = "INACTIVE"
)

// How close two item names have to be before the platform offers one as a match.
//
// Suggestions only: nothing here refuses a create or joins a request. A customer who
// means to open their own request opens it, and being shown demand they did not want
// costs them a glance - so this floor sits low, where it catches a typo or an extra
// word without needing to be right about it.
const (
	SuggestSimilarity float32 = 0.3

	// Enough to recognise the item among; past that it is a list nobody reads.
	maxSimilar int32 = 5
)

const uniqueViolation = "23505"

// outboxWaker lets the service nudge the relay the moment an event is committed, so a
// notification is not held for the poll interval when the broker is healthy.
type outboxWaker interface{ Wake() }

type Service struct {
	pool    *pgxpool.Pool
	queries *store.Queries
	relay   outboxWaker
}

func NewService(pool *pgxpool.Pool, relay outboxWaker) *Service {
	return &Service{pool: pool, queries: store.New(pool), relay: relay}
}

func (s *Service) wakeRelay() {
	if s.relay != nil {
		s.relay.Wake()
	}
}

type CreateInput struct {
	ItemName    string
	Description string
	Category    string
	Quantity    int32

	// The token subject of the customer acting. Notification-service is addressed by
	// global userId and never resolves an identity, so the producer supplies the one it
	// already holds - and it has to reach this far in, because the event is written
	// inside the same transaction as the request itself.
	ActorUserID uuid.UUID
}

// EnsureInput describes an item to find or open a request for. There is no quantity:
// nobody is being enrolled, which is the whole difference from CreateInput.
type EnsureInput struct {
	ItemName    string
	Description string
	Category    string
}

type ListFilter struct {
	ItemName string
	Category string
	Status   string
	Limit    int32
	Offset   int32
}

// Create opens a request and enrolls its creator as the first participant, in one
// transaction. R1 and R3 travel together: a customer creates a request because they
// want some quantity of the item, so a request never exists with nobody wanting it.
//
// Except when that item already has a request. Two requests for "Espresso Machine"
// split the number a seller bids against, which is the number the whole platform exists
// to build, so there is nothing for the second one to create - it is refused and handed
// the request that exists. Joining it is then the customer's own call, made through the
// participants endpoint: being quietly enrolled in a stranger's request because a name
// collided is not what anybody asked for.
//
// That holds whether or not anyone is on it. A request with no buyers is the one a join
// helps most - it is how the item becomes demand again - and it may already carry a
// seller's offer waiting for buyers to arrive.
//
// A merely similar name is not refused. Whether "Espresso Machine Pro" is the same
// product is a judgement this has only the spelling to make, so those reach the customer
// as suggestions while they type - see Similar - and they decide.
func (s *Service) Create(
	ctx context.Context, customerID uuid.UUID, in CreateInput,
) (store.PurchaseRequest, error) {
	var created store.PurchaseRequest

	err := s.inTx(ctx, func(q *store.Queries, tx pgx.Tx) error {
		// Before looking, so the answer cannot go stale between the look and the write:
		// two customers naming the same new item at once would otherwise both find
		// nothing and both create it.
		if err := q.LockItemName(ctx, in.ItemName); err != nil {
			return fmt.Errorf("locking item name: %w", err)
		}

		existing, err := q.FindRequestByItemName(ctx, in.ItemName)
		switch {
		case err == nil:
			return s.refuseExisting(ctx, q, existing, customerID)
		case !errors.Is(err, pgx.ErrNoRows):
			return fmt.Errorf("looking for an open request for the same item: %w", err)
		}

		request, err := q.CreateRequest(ctx, store.CreateRequestParams{
			ItemName:    in.ItemName,
			Description: in.Description,
			Category:    in.Category,
			// R1: the request has an owner from the moment it exists, and that is who
			// may later close it.
			CreatedBy: pgtype.UUID{Bytes: customerID, Valid: true},
		})
		if err != nil {
			return fmt.Errorf("creating request: %w", err)
		}

		if _, err := q.AddParticipant(ctx, store.AddParticipantParams{
			RequestID:  request.RequestID,
			CustomerID: customerID,
			Quantity:   in.Quantity,
		}); err != nil {
			return fmt.Errorf("adding creator as participant: %w", err)
		}

		created, err = q.RecalculateDemand(ctx, request.RequestID)
		if err != nil {
			return fmt.Errorf("recalculating demand: %w", err)
		}

		// Flow 1, step 8 - in the same transaction, so the request and the promise to
		// tell someone about it either both exist or neither does.
		return events.Enqueue(ctx, tx, events.KeyRequestJoined, events.Notification{
			UserID: in.ActorUserID,
			Type:   "REQUEST_JOINED",
			Title:  "Your request is open",
			Message: fmt.Sprintf(
				"Your request for %s is open. You are its first participant, wanting %d.",
				created.ItemName, in.Quantity),
		})
	})

	s.wakeRelay()
	return created, err
}

// EnsureForItem returns the request an item is carried by, opening one with nobody on it
// if there is none. It reports whether it had to create it.
//
// This is what lets a seller offer against an item nobody has asked for yet. Demand and
// supply do not have to arrive in that order: a seller who already holds the stock is
// worth letting speak first, and the request their offer needs is the thing that gives
// buyers somewhere to arrive. Such a request is INACTIVE, which is not a special case -
// INACTIVE is exactly what "no participants" is called, the same state a request reaches
// when its last customer leaves, and the same join revives either.
//
// It creates nothing when the item already has a request, for the same reason Create
// refuses one: two requests for the same item split the total the platform exists to
// pool. The seller's offer joins whatever demand is already there instead.
//
// No owner and no participants, so:
//   - created_by stays NULL. A seller is not a customer, cannot join, and must not be
//     recorded as the buyer who wanted this.
//   - no REQUEST_JOINED event. Nobody joined, and the seller learns the outcome in the
//     response to the offer they were making.
func (s *Service) EnsureForItem(
	ctx context.Context, in EnsureInput,
) (store.PurchaseRequest, bool, error) {
	var (
		request store.PurchaseRequest
		created bool
	)

	err := s.inTx(ctx, func(q *store.Queries, tx pgx.Tx) error {
		// The same lock Create takes, and it has to be the same one: a customer naming
		// an item at the instant a seller offers against it would otherwise have both
		// find nothing and both create a request for it.
		if err := q.LockItemName(ctx, in.ItemName); err != nil {
			return fmt.Errorf("locking item name: %w", err)
		}

		existing, err := q.FindRequestByItemName(ctx, in.ItemName)
		switch {
		case err == nil:
			request = existing
			return nil
		case !errors.Is(err, pgx.ErrNoRows):
			return fmt.Errorf("looking for a request for the same item: %w", err)
		}

		opened, err := q.CreateRequest(ctx, store.CreateRequestParams{
			ItemName:    in.ItemName,
			Description: in.Description,
			Category:    in.Category,
			// Left NULL deliberately - see the note above, and 000003, which made the
			// column nullable for exactly the case of a request with no buyer behind it.
			CreatedBy: pgtype.UUID{},
		})
		if err != nil {
			return fmt.Errorf("creating request: %w", err)
		}

		// Not skipped just because the answer is known: the status is derived from the
		// participant count by one query, and letting this path write 'INACTIVE' itself
		// would be a second place that decides it.
		request, err = q.RecalculateDemand(ctx, opened.RequestID)
		if err != nil {
			return fmt.Errorf("recalculating demand: %w", err)
		}
		created = true
		return nil
	})

	return request, created, err
}

// refuseExisting turns a name collision into the right refusal. A caller who is already
// in that request is told so plainly - "join this instead" is no use to somebody who
// already has - and everyone else is handed it to join.
func (s *Service) refuseExisting(
	ctx context.Context, q *store.Queries, existing store.PurchaseRequest, customerID uuid.UUID,
) error {
	_, err := q.GetParticipant(ctx, store.GetParticipantParams{
		RequestID:  existing.RequestID,
		CustomerID: customerID,
	})
	switch {
	case err == nil:
		return ErrAlreadyParticipant
	case !errors.Is(err, pgx.ErrNoRows):
		return fmt.Errorf("checking existing participation: %w", err)
	}
	return &RequestExistsError{Existing: existing}
}

// Similar backs the suggestions a customer sees while typing an item name. It is a
// plain read - the same projection browsing already serves - so it needs no token and
// takes no lock.
//
// Requests with nobody on them are included, and each row carries its status. One of
// those is often the most useful thing to suggest: an item a seller has already offered
// against is waiting for its first buyer, and joining is what makes it demand.
func (s *Service) Similar(ctx context.Context, itemName string, minScore float32) ([]store.FindSimilarRequestsRow, error) {
	return s.queries.FindSimilarRequests(ctx, store.FindSimilarRequestsParams{
		ItemName:    itemName,
		MinScore:    minScore,
		ResultLimit: maxSimilar,
	})
}

func (s *Service) Get(ctx context.Context, requestID uuid.UUID) (store.PurchaseRequest, error) {
	request, err := s.queries.GetRequest(ctx, requestID)
	if errors.Is(err, pgx.ErrNoRows) {
		return store.PurchaseRequest{}, ErrRequestNotFound
	}
	return request, err
}

func (s *Service) List(ctx context.Context, f ListFilter) ([]store.PurchaseRequest, error) {
	return s.queries.ListRequests(ctx, store.ListRequestsParams{
		ItemName:     optionalText(f.ItemName),
		Category:     optionalText(f.Category),
		Status:       optionalText(f.Status),
		ResultLimit:  f.Limit,
		ResultOffset: f.Offset,
	})
}

func (s *Service) ListForCustomer(ctx context.Context, customerID uuid.UUID) ([]store.PurchaseRequest, error) {
	return s.queries.ListRequestsByCustomer(ctx, customerID)
}

func (s *Service) ParticipantCustomerIDs(ctx context.Context, requestID uuid.UUID) ([]uuid.UUID, error) {
	if _, err := s.Get(ctx, requestID); err != nil {
		return nil, err
	}
	return s.queries.ListParticipantCustomerIDs(ctx, requestID)
}

// Join adds a customer to an open request (R2, R3).
func (s *Service) Join(
	ctx context.Context, requestID, customerID, actorUserID uuid.UUID, quantity int32,
) (store.PurchaseRequest, error) {
	joined, err := s.mutateParticipants(ctx, requestID,
		func(q *store.Queries, tx pgx.Tx) error {
			_, err := q.AddParticipant(ctx, store.AddParticipantParams{
				RequestID:  requestID,
				CustomerID: customerID,
				Quantity:   quantity,
			})

			// The unique constraint, not a prior read, is what makes a double join
			// impossible: two concurrent requests would both pass a read-then-write check.
			var pgErr *pgconn.PgError
			if errors.As(err, &pgErr) && pgErr.Code == uniqueViolation {
				return ErrAlreadyParticipant
			}
			return err
		},
		// Enqueued inside the same transaction, after the totals have been recomputed,
		// so the joiner is told the demand they actually became part of.
		func(updated store.PurchaseRequest) []events.Notification {
			return []events.Notification{{
				UserID: actorUserID,
				Type:   "REQUEST_JOINED",
				Title:  "You joined a request",
				Message: fmt.Sprintf(
					"You joined the request for %s, wanting %d. It now has %d customers wanting %d in total.",
					updated.ItemName, quantity, updated.TotalCustomers, updated.TotalQuantity),
			}}
		})

	s.wakeRelay()
	return joined, err
}

// UpdateQuantity changes the caller's own quantity. The caller is identified by the
// token, so one customer can never edit another's line.
func (s *Service) UpdateQuantity(ctx context.Context, requestID, customerID uuid.UUID, quantity int32) (store.PurchaseRequest, error) {
	return s.mutateParticipants(ctx, requestID, func(q *store.Queries, tx pgx.Tx) error {
		_, err := q.UpdateParticipantQuantity(ctx, store.UpdateParticipantQuantityParams{
			RequestID:  requestID,
			CustomerID: customerID,
			Quantity:   quantity,
		})
		if errors.Is(err, pgx.ErrNoRows) {
			return ErrNotParticipant
		}
		return err
	}, nil)
}

func (s *Service) Leave(ctx context.Context, requestID, customerID uuid.UUID) (store.PurchaseRequest, error) {
	return s.mutateParticipants(ctx, requestID, func(q *store.Queries, tx pgx.Tx) error {
		affected, err := q.DeleteParticipant(ctx, store.DeleteParticipantParams{
			RequestID:  requestID,
			CustomerID: customerID,
		})
		if err != nil {
			return err
		}
		if affected == 0 {
			return ErrNotParticipant
		}
		return nil
	}, nil)
}

// mutateParticipants runs a participant change and the demand recalculation it implies
// as one unit. The request row is locked first so concurrent joins on the same request
// serialize: without it two transactions could each recompute totals from a snapshot
// that misses the other's insert (R4).
func (s *Service) mutateParticipants(
	ctx context.Context,
	requestID uuid.UUID,
	mutate func(*store.Queries, pgx.Tx) error,
	notify func(store.PurchaseRequest) []events.Notification,
) (store.PurchaseRequest, error) {
	var updated store.PurchaseRequest

	err := s.inTx(ctx, func(q *store.Queries, tx pgx.Tx) error {
		// The lock still matters - it is what serializes concurrent joins so the totals
		// below cannot be computed from a snapshot missing one - but the row it returns
		// is no longer read: there is no status left to gate on.
		if _, err := q.LockRequest(ctx, requestID); errors.Is(err, pgx.ErrNoRows) {
			return ErrRequestNotFound
		} else if err != nil {
			return fmt.Errorf("locking request: %w", err)
		}
		// Deliberately ungated. An INACTIVE request is one nobody is on rather than one
		// that has ended, so joining it is how it comes back - refusing here would make
		// the last customer to leave the one who closed it for good.
		if err := mutate(q, tx); err != nil {
			return err
		}

		var err error
		updated, err = q.RecalculateDemand(ctx, requestID)
		if err != nil {
			return fmt.Errorf("recalculating demand: %w", err)
		}

		if notify == nil {
			return nil
		}
		return events.Enqueue(ctx, tx, events.KeyRequestJoined, notify(updated)...)
	})

	return updated, err
}

// Nothing sets a status. Close and SetStatus lived here: the owner ended the request
// for everyone, and Admin/Contact marked it approved. Both are gone - an owner is a
// participant like any other and leaves, and an approval decides a seller rather than
// the demand. What is left of either is RecalculateDemand deriving the status from who
// is still on the request.

// inTx hands the callback both the generated queries and the raw transaction: the
// domain writes go through sqlc, and the outbox insert needs the tx itself, but they
// have to be the same transaction for the outbox to mean anything.
func (s *Service) inTx(ctx context.Context, fn func(*store.Queries, pgx.Tx) error) error {
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return fmt.Errorf("beginning transaction: %w", err)
	}
	// No-op once the transaction commits; guarantees rollback on every early return.
	defer func() { _ = tx.Rollback(ctx) }()

	if err := fn(s.queries.WithTx(tx), tx); err != nil {
		return err
	}
	return tx.Commit(ctx)
}

// optionalText maps an empty filter value to SQL NULL, which the queries read as
// "no filter" rather than "match the empty string".
func optionalText(v string) pgtype.Text {
	if v == "" {
		return pgtype.Text{}
	}
	return pgtype.Text{String: v, Valid: true}
}

File: offer-service/app/service.py
"""Offer domain logic (spec section 11)."""

import uuid
from collections.abc import Sequence
from datetime import UTC, datetime

from sqlalchemy import func, select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from app.events import KEY_OFFER_CREATED, enqueue
from app.models import Offer, OfferImage, OfferStatus
from app.schemas import OfferCreate, OfferOut, OfferUpdate


class OfferNotFound(Exception):
    pass


class NotOfferOwner(Exception):
    """A seller may only touch their own offer."""


class OfferAlreadyMade(Exception):
    """The seller already has a live offer on this request.

    Answering the same demand twice is one proposal changed, not two - so it carries the
    offer that already exists, because "you cannot do this" without saying what to do
    instead leaves the seller nowhere to go. Updating it is PUT /api/offers/{offerId},
    which is the call this refusal points at.

    It carries the projection rather than the ORM instance on purpose. get_session rolls
    the session back when a handler raises, which expires every instance loaded from it,
    and the exception handler that renders this runs after the session has closed - so a
    live object would be detached and unreadable by the time anyone asked it anything.
    """

    def __init__(self, existing: Offer):
        super().__init__("this seller already has a live offer on this request")
        self.existing = OfferOut.of(existing)


class OfferNotPending(Exception):
    """Once an offer has been decided, its terms are frozen: an approved offer may
    already have contact permission granted against it, so letting the seller rewrite
    the price afterwards would change what the admin actually approved."""


# Nothing here refuses an offer on the strength of the request's status. A request is
# OPEN or INACTIVE and neither ends it, so there is no state in which demand has gone
# for good and an offer against it is a mistake.
#
# What does refuse one is the seller having answered this request already - see
# OfferAlreadyMade. That is a rule about the seller, not about the demand: other sellers
# still compete freely on the same request, and an approval does not close the door on
# the ones who come after.


async def create_offer(
    session: AsyncSession,
    seller_id: uuid.UUID,
    request_id: uuid.UUID,
    payload: OfferCreate,
    admin_user_ids: Sequence[uuid.UUID] = (),
    image: bytes | None = None,
    image_type: str = "",
) -> Offer:
    """R6: a new offer always starts PENDING. The caller cannot choose otherwise -
    status is not part of OfferCreate at all.

    request_id is passed rather than read off the payload, because the payload may not
    carry one: a seller who named an item instead had the request resolved - or opened -
    by request-service on the way in. Either way an offer is stored against a request
    that exists, which is all this layer needs to be true.

    A seller answers a request once: a second offer against demand they have already bid
    on is refused with the first one attached, to be updated instead. The read below is
    what makes that a clear 409 rather than a constraint violation, and the unique index
    behind it is what makes it true under a race the read cannot see.

    The NEW_OFFER event of flow 2 step 7 is written to the outbox in this same
    transaction, so the offer and the promise to tell the admins about it either both
    exist or neither does. admin_user_ids is empty when auth-service could not be
    reached, in which case there is simply nothing to announce.
    """
    existing = await live_offer_of(session, seller_id, request_id)
    if existing is not None:
        raise OfferAlreadyMade(existing)

    offer = Offer(
        seller_id=seller_id,
        request_id=request_id,
        available_quantity=payload.available_quantity,
        price_per_unit=payload.price_per_unit,
        currency=payload.currency,
        description=payload.description,
        status=OfferStatus.PENDING,
        # Both or neither: image_type is what every reader uses to decide whether a
        # picture exists, so bytes without it would be hidden and a type without bytes
        # would promise an image the endpoint cannot serve.
        image_type=image_type if image else "",
    )
    session.add(offer)

    # Flushed before the event is built so the offer's own columns are settled - the
    # message quotes the terms an admin is being asked to review. It is also what gives
    # the offer the id its picture is keyed by.
    await session.flush()

    # In this transaction rather than a call after it, so an offer never exists in a
    # state where its picture is still on its way - and a submission refused as a
    # duplicate leaves no orphaned image behind.
    if image:
        session.add(OfferImage(offer_id=offer.offer_id, image_data=image))

    enqueue(
        session,
        KEY_OFFER_CREATED,
        [
            {
                "userId": str(admin_id),
                "type": "NEW_OFFER",
                "title": "An offer is waiting for review",
                "message": (
                    f"A seller offered {offer.available_quantity} at "
                    f"{offer.price_per_unit} {offer.currency} per unit."
                ),
            }
            for admin_id in admin_user_ids
        ],
    )

    try:
        await session.commit()
    except IntegrityError as exc:
        # The index, not the read above, is what makes a double offer impossible: two
        # submissions arriving together would both find nothing and both insert.
        await session.rollback()
        existing = await live_offer_of(session, seller_id, request_id)
        if existing is None:
            raise
        raise OfferAlreadyMade(existing) from exc

    await session.refresh(offer)
    return offer


async def live_offer_of(
    session: AsyncSession, seller_id: uuid.UUID, request_id: uuid.UUID
) -> Offer | None:
    """The seller's standing offer on this request, if they have one.

    Live is PENDING or APPROVED - a proposal that still stands. A cancelled or rejected
    one is a record of a proposal that does not, and neither can be updated, so treating
    either as "you already offered" would leave the seller unable to offer and unable to
    update.
    """
    stmt = select(Offer).where(
        Offer.seller_id == seller_id,
        Offer.request_id == request_id,
        Offer.status.in_(OfferStatus.LIVE),
    )
    return (await session.scalars(stmt)).first()


async def count_live_offers(session: AsyncSession, request_id: uuid.UUID) -> int:
    """How many offers on this request still stand.

    PENDING or APPROVED - the same set that stops a seller offering twice. A cancelled or
    rejected offer is a record of a proposal rather than one, so it neither blocks a new
    offer nor keeps a request from going dormant.
    """
    stmt = (
        select(func.count())
        .select_from(Offer)
        .where(Offer.request_id == request_id, Offer.status.in_(OfferStatus.LIVE))
    )
    return int(await session.scalar(stmt) or 0)


async def get_offer(session: AsyncSession, offer_id: uuid.UUID) -> Offer:
    offer = await session.get(Offer, offer_id)
    if offer is None:
        raise OfferNotFound
    return offer


async def list_for_request(session: AsyncSession, request_id: uuid.UUID) -> Sequence[Offer]:
    stmt = select(Offer).where(Offer.request_id == request_id).order_by(Offer.created_at)
    return (await session.scalars(stmt)).all()


async def list_for_seller(session: AsyncSession, seller_id: uuid.UUID) -> Sequence[Offer]:
    stmt = select(Offer).where(Offer.seller_id == seller_id).order_by(Offer.created_at.desc())
    return (await session.scalars(stmt)).all()


async def list_pending(session: AsyncSession, limit: int, offset: int) -> Sequence[Offer]:
    stmt = (
        select(Offer)
        .where(Offer.status == OfferStatus.PENDING)
        .order_by(Offer.created_at)
        .limit(limit)
        .offset(offset)
    )
    return (await session.scalars(stmt)).all()


async def update_offer(
    session: AsyncSession,
    offer_id: uuid.UUID,
    seller_id: uuid.UUID,
    payload: OfferUpdate,
    image: bytes | None = None,
    image_type: str = "",
) -> Offer:
    offer = await _own_pending_offer(session, offer_id, seller_id)

    offer.available_quantity = payload.available_quantity
    offer.price_per_unit = payload.price_per_unit
    offer.currency = payload.currency
    offer.description = payload.description
    offer.updated_at = _now()

    # No image in the body means the picture is untouched, not removed. The seller is
    # editing terms in a form that shows the picture they already uploaded; re-sending
    # it on every price change would be the client's work to do for no reason, and
    # reading its absence as a delete would lose it by accident.
    if image:
        await _replace_image(session, offer, image, image_type)

    await session.commit()
    await session.refresh(offer)
    return offer


async def _replace_image(
    session: AsyncSession, offer: Offer, image: bytes, image_type: str
) -> None:
    """An offer shows one picture, so a new one replaces the old rather than joining it."""
    existing = await session.get(OfferImage, offer.offer_id)
    if existing is None:
        session.add(OfferImage(offer_id=offer.offer_id, image_data=image))
    else:
        existing.image_data = image
        existing.updated_at = _now()
    offer.image_type = image_type


async def offer_image(session: AsyncSession, offer_id: uuid.UUID) -> tuple[bytes, str, object]:
    """The stored picture, the type to serve it as, and when it last changed.

    Raises OfferNotFound when the offer has no picture - or does not exist. One answer
    for both: there is nothing at that URL either way, and distinguishing them would
    tell a caller which offer ids are real.
    """
    offer = await session.get(Offer, offer_id)
    if offer is None or not offer.image_type:
        raise OfferNotFound

    stored = await session.get(OfferImage, offer_id)
    if stored is None:
        raise OfferNotFound
    return stored.image_data, offer.image_type, stored.updated_at


async def cancel_offer(session: AsyncSession, offer_id: uuid.UUID, seller_id: uuid.UUID) -> Offer:
    """Cancelling is a status change, not a delete: the offer stays as a record of what
    was proposed, which the audit history in Admin/Contact refers back to."""
    offer = await _own_pending_offer(session, offer_id, seller_id)

    offer.status = OfferStatus.CANCELLED
    offer.updated_at = _now()

    await session.commit()
    await session.refresh(offer)
    return offer


async def set_status(session: AsyncSession, offer_id: uuid.UUID, new_status: str) -> Offer:
    """Records an admin decision relayed by Admin/Contact. Only a PENDING offer can be
    decided, so a second approval - or approving something already cancelled - is
    refused rather than silently overwriting the first outcome."""
    offer = await get_offer(session, offer_id)
    if offer.status != OfferStatus.PENDING:
        raise OfferNotPending

    offer.status = new_status
    offer.updated_at = _now()

    await session.commit()
    await session.refresh(offer)
    return offer


async def _own_pending_offer(
    session: AsyncSession, offer_id: uuid.UUID, seller_id: uuid.UUID
) -> Offer:
    offer = await get_offer(session, offer_id)
    if offer.seller_id != seller_id:
        # Deliberately distinct from OfferNotFound: the caller is authenticated and the
        # offer exists, they simply do not own it.
        raise NotOfferOwner
    if offer.status != OfferStatus.PENDING:
        raise OfferNotPending
    return offer


def _now() -> datetime:
    return datetime.now(UTC)
